tests/firmware_tests.py

```diff
@@ -0,0 +1,346 @@
+        message_mock.topic = "v1/devices/me/attributes_update"
+        payload_dict = {
+            "fw_version": "v1",
+            "fw_title": "TestFirmware",
+            "fw_size": 900
+        }
+        message_mock.payload = orjson.dumps(payload_dict)
+
+        client._on_decoded_message({}, message_mock)
+        client.stopped = True
+
+        client._messages_rate_limit.increase_rate_limit_counter.assert_called_once()
+
+        calls = [args[0] for args, kwargs in mock_log_debug.call_args_list]
+        self.assertTrue(any("Firmware is not the same" in call for call in calls),
+                        f"Expected log.debug call with 'Firmware is not the same', got: {calls}")
+
+        self.assertEqual(client.firmware_data, b"")
+        self.assertEqual(client._TBDeviceMqttClient__current_chunk, 0)
+        self.assertEqual(client.current_firmware_info[FW_STATE_ATTR], "DOWNLOADING")
+
+        client.send_telemetry.assert_called_once_with(client.current_firmware_info)
+
+        sleep_called = any(args and (args[0] == 1 or args[0] == 1.0) for args, kwargs in mock_sleep.call_args_list)
+        self.assertTrue(sleep_called, f"sleep(1) was not called, calls: {mock_sleep.call_args_list}")
+
+        self.assertEqual(client._TBDeviceMqttClient__firmware_request_id, 1)
+        self.assertEqual(client._TBDeviceMqttClient__target_firmware_length, 900)
+        self.assertEqual(client._TBDeviceMqttClient__chunk_count, ceil(900 / 128))
+        client._TBDeviceMqttClient__get_firmware.assert_called_once()
+
+
+class TestTBDeviceMqttClientOnConnect(unittest.TestCase):
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_success(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+        client._subscribe_to_topic = MagicMock()
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=0)
+
+        self.assertTrue(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_not_called()
+
+        expected_sub_calls = [
+            call('v1/devices/me/attributes', qos=client.quality_of_service),
+            call('v1/devices/me/attributes/response/+', qos=client.quality_of_service),
+            call('v1/devices/me/rpc/request/+', qos=client.quality_of_service),
+            call('v1/devices/me/rpc/response/+', qos=client.quality_of_service),
+        ]
+        client._subscribe_to_topic.assert_has_calls(expected_sub_calls, any_order=False)
+
+        self.assertTrue(client._TBDeviceMqttClient__request_service_configuration_required)
+
+    @patch('tb_device_mqtt.log')
+    def test_on_connect_fail_known_code(self, mock_logger):
+        client = TBDeviceMqttClient("<THINGSBOARD_HOST>", 1883, "<ACCESS_TOKEN>")
+
+        client._on_connect(client=None, userdata=None, flags=None, result_code=1)
+
+        self.assertFalse(client._TBDeviceMqttClient__is_connected)
+        mock_logger.error.assert_called_once_with(
```

Comment:
Using a logger message as a result of the test is not correct.
A test can be affected by simply changing the logger message but tested functionality will still be correct.